ziggurat-core-geoip/src/geoip.rs

```diff
@@ -1,130 +1,36 @@
-//use ipinfo::{IpInfo, IpInfoConfig};   // TODO(asmie): implement IpInfo provider
-use ip2location::{Record, DB};
-use ipgeolocate::{Locator, Service};
+use std::net::IpAddr;
 
-/// List of supported geoip providers.
-pub enum Provider {
-    IpApiCom,
-    IpApiCo,
-    IpInfo,
-    Ip2Location,
+use async_trait::async_trait;
+
+/// Every provider need to implement this trait.
+#[async_trait]
+pub trait GeoIPService {
+    /// Lookup the IP address and return the GeoIPInfo.
+    async fn lookup(&self, ip: IpAddr) -> Result<GeoIPInfo, String>;
 }
 
-/// Geoip provider service configuration.
-pub struct GeoIPService {
-    /// Geoip provider. Look at `Provider` enum for supported providers.
-    pub provider: Provider,
-    /// API key (paid or free) for online providers. For offline providers put here path to database file.
-    pub api_key: String,
+/// Geographical location of an IP address.
+#[derive(Copy, Clone, PartialEq)]
+pub struct Location {
+    pub latitude: f64,
+    pub longitude: f64,
 }
 
 /// IP information
-pub struct IPGeoInfo {
+#[derive(Clone)]
+pub struct GeoIPInfo {
     /// IP address
-    pub ip: String,
+    pub ip: IpAddr,
     /// Country name (long name)
     pub country: Option<String>,
     /// City name
     pub city: Option<String>,
-    /// IP address latitude
+    /// Latitude
     pub latitude: Option<f64>,
-    /// IP address longitude
+    /// Longitude
     pub longitude: Option<f64>,
```

Comment:
Should we use the new `Location` struct here? 😃

Reply:
For now I left two values as each provider splits them (for some reason).
I'm leaving it on my todo list when I'll learn more about the real world usages :)

Reply:
aha, now I was confused by my comment... 😅

I see now the `Location` struct had been removed even before I left this comment - I may have done a review on an open tab which didn't include the latest change with the struct removed